JuliaStream.jl/src/ThreadedStream.jl

```diff
@@ -63,12 +63,46 @@ function nstream!(data::VectorData{T}, _) where {T}
   end
 end
 
+# Threads.@threads/Threads.@spawn doesn't support OpenMP's firstprivate, etc
+function static_par_ranged(f::Function, range::Int, n::Int)
+  stride = range ÷ n
+  rem = range % n
+  strides = map(0:n) do i
+    width = stride + (i < rem ? 1 : 0)
+    offset = i < rem ? (stride + 1) * i : ((stride + 1) * rem) + (stride * (i - rem))
+    (offset, width)
+  end
+  ccall(:jl_enter_threaded_region, Cvoid, ())
+  try
+    foreach(wait, map(1:n) do group
+      (offset, size) = strides[group]
+      task = Task(() -> f(group, offset+1, offset+size))
+      task.sticky = true
+      ccall(:jl_set_task_tid, Cvoid, (Any, Cint), task, group-1) # ccall, so 0-based for group
+      schedule(task)
+    end)
+  finally
+    ccall(:jl_exit_threaded_region, Cvoid, ())
+  end
+end
+
 function dot(data::VectorData{T}, _) where {T}
-  partial = zeros(T, Threads.nthreads())
-  Threads.@threads for i = 1:data.size
-    @inbounds partial[Threads.threadid()] += data.a[i] * data.b[i]
+  partial = Vector{T}(undef,  Threads.nthreads())
+  static_par_ranged(data.size, Threads.nthreads()) do group, startidx, endidx
+    acc = zero(T)
+    @fastmath for i = startidx:endidx
+      @inbounds acc += data.a[i] * data.b[i]
+    end
+    @inbounds partial[group] = acc
   end
   return sum(partial)
+  # This doesn't do well on aarch64 because of the excessive Threads.threadid() ccall 
+  # and inhibited vectorisation from the lack of @fastmath 
+  # partial = zeros(T, Threads.nthreads())
+  # Threads.@threads for i = 1:data.size
```

Comment:
Normally I would do a:

```
@threads for tid in 1:nthreads()
    region = ....
    acc = 0.0
    @simd for i in region
       acc += ...
    end
    partial[i] = acc
end
```

Or using @tkf FLoops.jl/Folds

Reply:
Is that semantically equivalent to the `static_par_ranged`?  (minus the `@simd` bit)

Reply:
I think so. Basically that's the canonical example of saying, I want one iteration per thread. Then you have to do the splitting manually, but you also don't have to reimplement the lower-level details xD

Reply:
https://github.com/JuliaFolds/FLoops.jl#parallel-loop is an alternative

Reply:
Yeah, it'd be better to use `@threads for i 1:nthreads()` than touching internals like `task.sticky = true` and `jl_set_task_tid`.

The main difference to FLoops.jl is that it lets the Julia runtime dynamic load-balancing. It doesn't matter if you only micro-benchmarking a parallel dot product but static scheduling like `@threads for` and `static_par_ranged` would fail when used inside other parallel functions (it has to wait for all worker threads a re available). So, _if_ the aim is to benchmark practical and composable pieces of routines in Julia, it may be better to use dynamic scheduling. On the other hand, if you want to benchmark what you can squeeze out from the Julia runtime at the moment, static scheduling like `@threads for` and `static_par_ranged` may be fine.

Another difference to FLoops.jl is that, by default, it uses divide-and-conquer (DAC) scheduling (`@spawn`s and `wait`s) while `@threads for` and `static_par_ranged`  use sequential scheduling. I find that DAC (in Julia) sometimes helps performance since it has larger parallelism (in the sense of work/span ratio), especially when there are a lot of cores.